proxy/proxy/client.go

```diff
@@ -292,33 +240,126 @@ func isExpired(cfg *tls.Config) bool {
 	return time.Now().After(cfg.Certificates[0].Leaf.NotAfter)
 }
 
-func (c *Client) cachedCfg(instance string) (string, *tls.Config, string) {
+// startRefresh kicks off a refreshCfg asynchronously, returning a channel that is closed
+// when it's complete and updates the cacheEntry when complete. This function should only
+// be called from the scope of "cachedCfg", which controls the logic around throttling refreshes.
+func (c *Client) startRefresh(instance string, refreshCfgBuffer time.Duration) chan bool {
+	done := make(chan bool)
+	go func() {
+		defer close(done)
+		addr, cfg, ver, err := c.refreshCfg(instance)
+
+		c.cacheL.Lock()
+		old := c.cfgCache[instance]
+		// if we failed to refresh cfg do not throw out potentially valid one
+		if err != nil && !isExpired(old.cfg) {
+			logging.Errorf("failed to refresh the ephemeral certificate for %s, returning previous cert instead: %v", instance, err)
+			addr, cfg, ver, err = old.addr, old.cfg, old.version, old.err
+		}
+		c.cfgCache[instance] = cacheEntry{
+			lastRefreshed: time.Now(),
+			err:           err,
+			addr:          addr,
+			version:       ver,
+			cfg:           cfg,
+			done:          done,
+		}
+		c.cacheL.Unlock()
+
+		certExpiration := cfg.Certificates[0].Leaf.NotAfter
+		now := time.Now()
+		timeToRefresh := certExpiration.Sub(now) - refreshCfgBuffer
+		if timeToRefresh <= 0 {
+			// If a new certificate expires before our buffer has expired,
+			// we should wait a bit and schedule a new refresh to much closer to the expiration's date
+			// TODO: This situation probably only occurs for Issue #622 - when the oauth2 token isn't refreshed before the cert is
+			timeToRefresh = certExpiration.Sub(now) - (5 * time.Second)
+			logging.Errorf("new ephemeral certificate expires sooner than expected (adjusting refresh time to compensate): current time: %v, certificate expires: %v", now, certExpiration)
+		}
+		go c.refreshCertAfter(instance, timeToRefresh)
+	}()
+	return done
+}
+
+// isValid returns true if the cacheEntry is still useable
+func isValid(c cacheEntry) bool {
+	// if the cfg is unset or if there was an error returned during the last attempt
+	return !(c.cfg == nil || c.err != nil)
+}
+
+func needsRefresh(e cacheEntry, refreshCfgBuffer time.Duration) bool {
+	if e.done == nil { // no refresh started
+		return true
+	} else if !isValid(e) || time.Since(e.lastRefreshed) < refreshCfgBuffer {
+		select {
+		default: // last refresh completed, so it's time for a new one
+			return true
+		case <-e.done: // new refresh already started, so we can wait on that
+			return false
+		}
+	}
+	return false
+}
+
+func (c *Client) cachedCfg(instance string) (string, *tls.Config, string, error) {
 	c.cacheL.RLock()
-	ret, ok := c.cfgCache[instance]
+
+	throttle := c.RefreshCfgThrottle
```

Comment:
tbh, I'm not sure. Both fields are under the `refreshCfgL` mutex, which makes me think they might be projected by it. However I'm now realizing they both are exported, which isn't great since the mutex isn't. 

They were previously inside a critical section, so I attempted to maintain the same here.

Reply:
Based on internal usage in this file, I see no other writes to these fields. I suspect these fields are intended to be supplied when the `Client` is created and not modified later. But since they're exported, I suppose it's possible they could be changed later.

Reply:
Will play it safe and leave it for now - I don't think it's a significant performance difference.